src/main.rs

```diff
@@ -317,36 +321,70 @@ async fn entry(jobserver_client: LazyJobserverClient) -> Result<()> {
     let mut uithread = UIThread::new(!opts.no_confirm);
 
-    // Launch target detection
-    let desired_targets = get_desired_targets(&opts.targets);
+    let (install_path, metadata, temp_dir) = block_in_place(|| -> Result<_> {
+        // Compute install directory
+        let (install_path, custom_install_path) = get_install_path(opts.install_path.as_deref());
+        let install_path = install_path.ok_or_else(|| {
+            error!("No viable install path found of specified, try `--install-path`");
+            miette!("No install path found or specified")
+        })?;
+        fs::create_dir_all(&install_path).map_err(BinstallError::Io)?;
+        debug!("Using install path: {}", install_path.display());
+
+        // Load metadata
+        let metadata = if !custom_install_path {
+            debug!("Reading binstall/crates-v1.json");
+            Some(metafiles::binstall_v1::Records::load()?)
+        } else {
+            None
+        };
 
-    // Compute install directory
-    let (install_path, custom_install_path) = get_install_path(opts.install_path.as_deref());
-    let install_path = install_path.ok_or_else(|| {
-        error!("No viable install path found of specified, try `--install-path`");
-        miette!("No install path found or specified")
+        // Create a temporary directory for downloads etc.
+        //
+        // Put all binaries to a temporary directory under `dst` first, catching
+        // some failure modes (e.g., out of space) before touching the existing
+        // binaries. This directory will get cleaned up via RAII.
+        let temp_dir = tempfile::Builder::new()
+            .prefix("cargo-binstall")
+            .tempdir_in(&install_path)
+            .map_err(BinstallError::from)
+            .wrap_err("Creating a temporary directory failed.")?;
+
+        Ok((install_path, metadata, temp_dir))
     })?;
-    fs::create_dir_all(&install_path).map_err(BinstallError::Io)?;
-    debug!("Using install path: {}", install_path.display());
-
-    // Create a temporary directory for downloads etc.
-    //
-    // Put all binaries to a temporary directory under `dst` first, catching
-    // some failure modes (e.g., out of space) before touching the existing
-    // binaries. This directory will get cleaned up via RAII.
-    let temp_dir = tempfile::Builder::new()
-        .prefix("cargo-binstall")
-        .tempdir_in(&install_path)
-        .map_err(BinstallError::from)
-        .wrap_err("Creating a temporary directory failed.")?;
+
+    // Remove installed crates
+    let crate_names = CrateName::dedup(crate_names).filter(|crate_name| {
+        if opts.force {
+            true
+        } else if let Some(records) = &metadata {
+            let keep = if let Some(version_req) = &crate_name.version_req {
+                records
+                    .get(&crate_name.name)
+                    .map(|metadata| !version_req.is_latest_compatible(&metadata.current_version))
```

Comment:
I don't think we can determine this purely from the arguments and the metafile: at some point we must go to the index to figure out which versions are available, and what will actually be installed, and whether that's the same as what's already there.

Reply:
Yeah, I can modify the `resolve` function to take in its `curr_version` and decides whether to update or not.

Reply:
I think that would be the most accurate. We can still do an early check here if the version_req is a `=` req and we've already got it installed, just to save on a resolve.

Reply:
Done!

`binstall::resolve` would skip download if the `curr_version == new_version` and just return `Resolution::AlreadyUpToDate`.

The `binstall::install` module needs a refactor though to expose the underlying functions instead of taking `Resolution`.